source/common/grpc/async_client_manager_impl.cc

```diff
@@ -33,6 +33,8 @@ AsyncClientFactoryImpl::AsyncClientFactoryImpl(Upstream::ClusterManager& cm,
   if (skip_cluster_check) {
     return;
   }
+  ENVOY_BUG(Thread::MainThread::isMainThread(),
+            "async client factory cluster checks should only be performed on the main thread");
```

Comment:
We require testing for ENVOY_BUG. Given that the behavior after hitting this is undefined I think this should be a normal ASSERT but I will defer to @asraa as the arbiter of the new rules. :)

Reply:
I was trying to use a mechanism that wouldn't break production use cases and would only fail in tests/dev. However, I'm not familiar with the style here for these kind of checks.

Also, an alternative would be to put the ENVOY_BUG/ASSERT in `ClusterManager::clusters()`, since that's the method that actually breaks.

Reply:
All of these checks currently use ASSERT so I would just switch to that.

Reply:
Undefined behavior after rules out ENVOY_BUG. It looks like this is a documentation check for a required condition that's unsafe if violated, so ASSERT SGTM

Can you please update the PR description?